### ISSUE TITLE
Consider vectorising LANLstar calculations

### ISSUE THREAD

**Issue by aaronhendry** [spacepy/spacepy#472]
The current calculation of LANLstar values for a large number of input points/epochs is rather more inefficient than it needs to be. Specifically, the for-loop in the _LANLcommon, from LANLstar.py:

https://github.com/spacepy/spacepy/blob/76c5a35c688e7876292ce05f3bf833a793d77fd6/spacepy/LANLstar.py#L144-L148

The actual neural-network calculation that takes place under the ffnet hood in this loop is a simple series of matrix/vector operations, but there is a significant level of overhead involved in each call to ffnet. It is possible to extract the various weights and biases needed to manually calculate the ANN value from the ffnet pickle object, and then convert the for-loop above into a series of numpy vector operations. This results in significantly increased throughput for the LANLstar functions -- it also has the added benefit of removing the soft-dependency on ffnet+networkx.

I have a working proof-of-concept for this idea, which accurately reproduces the results from ffnet, however unfortunately it is currently written in MATLAB script (for various reasons it was easier for me to do this at the time in MATLAB than modify SpacePy to do this). It shouldn't be too arduous to convert it to work within SpacePy, but I guess the first question is whether this is even something that the SpacePy maintainers are interested in?

**Comment by jtniehof** [comment]
Removing the dependency on ffnet? That alone has me interested!

**Comment by drsteve** [comment]
Hi @aaronhendry 
Thanks for that. Yes, we're interested and porting your MatLab proof-of-concept will be much faster than starting the process from scratch, so thanks for your effort!

Just as a historical note, the dependence on `ffnet` is because that was (back in '09) pretty much the best package for training feed-forward neural networks in Python, and more importantly because very few people in space physics used Python (at the time) `ffnet` would export the trained network to a fortran subroutine so it was easy to distribute and call from other languages.
So the "least work" to include it in SpacePy once we became well-established and folks were actually using Python, was to just use ffnet to load/call.

Since there's no external development of LANLstar any more, and any new version is likely to be done from scratch using a more modern framework, it's high time that we simplified how it's included in SpacePy. Just pulling out the weights and solving directly will be faster, easier to maintain, and will simplify distribution (`networkx` is easy as a dependency, but `ffnet` isn't always easy)

If you're happy to share your code openly, then maybe just drop it into a gist ([gist.github.com](https://gist.github.com)) and paste the link here. If you'd rather just work with us (or have us give it a first pass) to start with then I'd still suggest dropping the code into a gist, but make it private and email me the link. I'll pass it to the other devs.

**Comment by aaronhendry** [comment]
Hi @drsteve, @jtniehof

I compiled what I've done into a series of gists. There are a few pre-processing steps that I did that are important, so I included those here as well. The first step is to convert the ffnet .net files to a more useful format -- for me this was a series of MATLAB matrices, but I actually used Python for the conversion, as I needed to use ffnet to load the .net files. This conversion only needs to be done once, so there is no dependence on ffnet once the files have been converted. I tried to document the code so it should be fairly self-explanatory, but I'm happy to try explain anything I've done if it's a bit too opaque.

Conversion code:
https://gist.github.com/aaronhendry/932f5a21d4d9fb6a2250bbc8d93a3026

Bash script to convert all .net files
https://gist.github.com/aaronhendry/25fc492e1253fb199ceed0d0e7833258

Now is the actual calculation, which is actually very simple. :
https://gist.github.com/aaronhendry/eb3b1144f1aafd1c992956c97e5d4752

Finally, I duplicated the test from SpacePy to make sure that the results work for all of the models (although I didn't test it for RAMSCB, as I had no use for this model and it takes the arguments in a slightly different fashion -- I didn't bother writing an edge case for this).
https://gist.github.com/aaronhendry/8d76576b92e5ba74825e6b5ba090c605

The output of the test:
```MATLAB
Model: OPDYN
	Expected: 4.7171
	Calculated: 4.7172
Model: OPQUIET
	Expected: 4.6673
	Calculated: 4.6673
Model: T01QUIET
	Expected: 4.8427
	Calculated: 4.8427
Model: T01STORM
	Expected: 4.8669
	Calculated: 4.8670
Model: T05
	Expected: 4.7174
	Calculated: 4.7175
Model: T89
	Expected: 4.5187
	Calculated: 4.5187
Model: T96
	Expected: 4.6439
	Calculated: 4.6439
```

There is a 0.0001 difference between the expected and calculated for a couple of them, but I assume this is okay.
